### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [jank-lang/jank#114] [Feature Requests & Ideas]
Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects (including many projects like compilers and interpreters)  - the results are available in the [awesome-pgo](https://github.com/zamazan4ik/awesome-pgo/) repo. Since PGO has helped in many cases, I think it would be a good idea to try optimizing Jank tools by applying PGO to them.

I can suggest the following things to do:

* We need to perform PGO benchmarks on the Jank ecosystem (like the compiler and the LSP server). If it shows improvements - add a note about possible improvements to the documentation. Providing an easier way (e.g. a build option) to build scripts to build Jank with PGO can be useful for the end-users too.
* It will be a good idea to integrate building with PGO into your CI pipelines. In this case, users will get already PGO-optimized binaries from the website (if you are going to provide prebuilt binaries, of course)

[Here](https://github.com/zamazan4ik/awesome-pgo/) you can find different materials about PGO: benchmarks in different software, examples of how PGO is already integrated with different projects, PGO support in multiple C++ compilers, and some PGO-related pieces of advice.

After PGO, I suggest evaluating the LLVM BOLT optimizer - it can give more aggressive optimizations even after PGO. However, starting with regular PGO will be easier to do.

I would be happy to answer all your questions about PGO!

**Comment by jeaye** [comment]
Hi! Apologies for the very long turnaround time. Since we weren't using Discussions much, any discussions made in the past year or two tended to get lost in my email.

I'm interested in utilizing PGO for jank, in a continuous way, to benefit our pre-compiled binaries. I'm not yet focusing much on performance, since I'm still working on parity with Clojure, but this will become a higher priority later this year.

I appreciate the offer for answering questions; I intend to take you up on that. It's my intention for jank to be the fastest Clojure dialect around, but we have a long way to go to get there. Every little bit will help.
